test/replication-luatest/gh_10561_anon_replicas_dont_participate_in_elections_test.lua

```diff
@@ -0,0 +1,82 @@
+local t = require('luatest')
+local cluster = require('luatest.replica_set')
+local fio = require('fio')
+local server = require('luatest.server')
+
+local g = t.group('replica-dont-participate-in-elections')
+--
+-- gh-10561:
+-- Anonymous replica don't participate in elections.
+--
+local wait_timeout = 10
+
+g.before_all(function(cg)
+    cg.cluster = cluster:new({})
+    cg.master = cg.cluster:build_and_add_server{alias = 'master'}
+    cg.replica_cfg = {
+        replication = server.build_listen_uri('master', cg.cluster.id),
+        -- Set manually for better test readability
+        election_mode = "off",
+        replication_anon = true,
+        read_only = true,
+    }
+    cg.replica = cg.cluster:build_and_add_server{
+        alias = 'replica',
+        box_cfg = cg.replica_cfg,
+    }
+    cg.cluster:start()
+    cg.master:exec(function()
+        box.ctl.promote()
+    end)
+    cg.replica:assert_follows_upstream(cg.master:get_instance_id())
+end)
+
+g.after_all(function(cg)
+    cg.cluster:drop()
+end)
+
+local err_msg_replication_anon =
+    "Incorrect value for option 'replication_anon': the value " ..
+    "may be set to true only when 'election_mode' is set to 'off'"
+local err_msg_election_mode =
+    "Incorrect value for option 'election_mode': the value may " ..
+    "only be set to 'off' when 'replication_anon' is set to true"
+
+for _, mode in ipairs({'candidate', 'manual', 'voter'}) do
+    g["test_anon_replica_startup_with_election_mode_" .. mode] = function(cg)
+        cg.replica:restart({box_cfg = {
+            replication = server.build_listen_uri('master', cg.cluster.id),
+            election_mode = mode,
+            replication_anon = true,
+            read_only = true,
+        }}, {wait_until_ready = false})
+        local log = fio.pathjoin(cg.replica.workdir, cg.replica.alias..'.log')
+        t.helpers.retrying({timeout = wait_timeout}, function()
+            local err_msg = err_msg_replication_anon
+            local ok = cg.replica:grep_log(err_msg, nil, {filename = log})
+            err_msg = err_msg_election_mode
+            ok = ok or cg.replica:grep_log(err_msg, nil, {filename = log})
```

Comment:
Why error msg can differ? AFAICS, only `err_msg_replication_anon` can be thrown here.

Applied the following diff. The test passes:

```diff
diff --git a/test/replication-luatest/gh_10561_anon_replicas_dont_participate_in_elections_test.lua b/test/replication-luatest/gh_10561_anon_replicas_dont_participate_in_elections_test.lua
index 236111b4ad..410d9b996f 100644
--- a/test/replication-luatest/gh_10561_anon_replicas_dont_participate_in_elections_test.lua
+++ b/test/replication-luatest/gh_10561_anon_replicas_dont_participate_in_elections_test.lua
@@ -53,10 +53,8 @@ for _, mode in ipairs({'candidate', 'manual', 'voter'}) do
         local log = fio.pathjoin(cg.replica.workdir, cg.replica.alias..'.log')
         t.helpers.retrying({timeout = wait_timeout}, function()
             local err_msg = err_msg_replication_anon
-            local ok = cg.replica:grep_log(err_msg, nil, {filename = log})
-            err_msg = err_msg_election_mode
-            ok = ok or cg.replica:grep_log(err_msg, nil, {filename = log})
-            t.assert(ok)
+            local msg = cg.replica:grep_log(err_msg, nil, {filename = log})
+            t.assert_not_equals(msg, nil)
         end)
         cg.replica:restart({box_cfg = cg.replica_cfg})
     end
```

Reply:
I reasoned that the order of checks is not a strict requirement, so theoretically it could change someday, and this is not bad behavior that we would like to eliminate with this test, but ok. Fixed.